torch/_inductor/codegen/cpp.py

```diff
@@ -2592,6 +2576,13 @@ def reduction_combine_vec(
                 # When combining intermediate accumulators we have a Welford<T> struct
                 mean, m2, weight = reduction_project(reduction_type, next_value)
             return f"welford_combine({var}, {{{mean}, {m2}, {weight}}})"
+        elif reduction_type in ("argmin", "argmax"):
+            if index is not None:
+                assert horizontal_reduction is not None
+                horizontal_str = "true" if horizontal_reduction else "false"
```

Comment:
#### Why we need `horizontal_reduction` for `argmin/argmax`
Summary: because we need horizontal_reduction` to know the `index` during reduce. And I put hthe logic for compute index into C++ functions `argmax_combine_vec`.
For this case:
```
    def test_argmax_argmin2(self):
        def fn(x):
            return (
                aten.argmax(x, 0),
                aten.argmin(x, 0),
                aten.argmax(x, 1),
                aten.argmin(x, 1),
            )

        self.common(fn, (torch.randn([144, 144]),))
```
The generated code are always load with contiguous dimension: (please note the `x0` and `x1` are different dim in the input tensor (torch.randn[144,144]) in below code)

So for `aten.argmax(x, 1)`, we are also compute the argmax at the contiguous dimension and we can use **x1 + offset** to represent the index
```
            for(long x0=static_cast<long>(0L); x0<static_cast<long>(144L); **x0+=static_cast<long>(1L)**)
            {
                {
                    IndexValue<float> tmp_acc0 = IndexValue<float>{0, -std::numeric_limits<float>::infinity()};
                    IndexValueVec<float, 16> tmp_acc0_vec = IndexValueVec<float, 16>(-std::numeric_limits<float>::infinity());
                    IndexValue<float> tmp_acc1 = IndexValue<float>{0, std::numeric_limits<float>::infinity()};
                    IndexValueVec<float, 16> tmp_acc1_vec = IndexValueVec<float, 16>(std::numeric_limits<float>::infinity());
                    for(long x1=static_cast<long>(0L); x1<static_cast<long>(144L); x1+=static_cast<long>(16L))
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + **static_cast<long>(x1 + (144L*x0)), 16)**;
                        tmp_acc0_vec = argmax_combine_vec(tmp_acc0_vec, tmp0, x1, true);
                        tmp_acc1_vec = argmin_combine_vec(tmp_acc1_vec, tmp0, x1, true);
                    }
                    tmp_acc0 = argmax_combine(tmp_acc0, argmax_vec_reduce_all(tmp_acc0_vec));
                    out_ptr2[static_cast<long>(x0)] = static_cast<int64_t>(tmp_acc0.index);
                    tmp_acc1 = argmin_combine(tmp_acc1, argmin_vec_reduce_all(tmp_acc1_vec));
                    out_ptr3[static_cast<long>(x0)] = static_cast<int64_t>(tmp_acc1.index);
                }
            }
```

And for `aten.argmax(x, 0)`
```
            for(long x0=static_cast<long>(0L); x0<static_cast<long>(144L); x0+=**static_cast<long>(16L)**)
            {
                {
                    IndexValue<float> tmp_acc0 = IndexValue<float>{0, -std::numeric_limits<float>::infinity()};
                    IndexValueVec<float, 16> tmp_acc0_vec = IndexValueVec<float, 16>(-std::numeric_limits<float>::infinity());
                    IndexValue<float> tmp_acc1 = IndexValue<float>{0, std::numeric_limits<float>::infinity()};
                    IndexValueVec<float, 16> tmp_acc1_vec = IndexValueVec<float, 16>(std::numeric_limits<float>::infinity());
                    for(long x1=static_cast<long>(0L); x1<static_cast<long>(144L); **x1+=static_cast<long>(1L)**)
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(x0 + (144L*x1)), 16);
                        tmp_acc0_vec = argmax_combine_vec(tmp_acc0_vec, tmp0, x1, false);
                        tmp_acc1_vec = argmin_combine_vec(tmp_acc1_vec, tmp0, x1, false);
                    }
                    tmp_acc0_vec.store(out_ptr0 + static_cast<long>(x0), 16);
                    tmp_acc1_vec.store(out_ptr1 + static_cast<long>(x0), 16);
                }
            }
        }
```
 we should use **x1** to represent the index
#### Why we do not need it for other reduction types:?
I think it is because other types like `sum/mean/max/` do not need to record the index

Reply:
For the second scenario `aten.argmax(x, 0)`, should we pass `x0` to the `argmax_combine_vec` instead of `x1`?
```c++
                        auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(x0 + (144L*x1)), 16);
                        tmp_acc0_vec = argmax_combine_vec(tmp_acc0_vec, tmp0, x1, false);
                        tmp_acc1_vec = argmin_combine_vec(tmp_acc1_vec, tmp0, x1, false);
```

Reply:
For the second scenario, the `x0` and `x1` is **switched** (compared with the first scenario). `x0` is second dimension (the contiguous dimension) and the `x1` is the first dimension.
`aten.argmax(x, 0)`  should return the max idx at the first dimension. So `x1` is correct.

Reply:
For the second scenario `aten.argmax(x, 0)`, the `x0, x1` is switched. `x0` is the second dimension (the contiguous dimension).  `aten.argmax(x, 0)` reduced on the **0** dimension which equal to the non-contiguous dimension `x1` in the generated code. So what we should to record is the max idx in `x1`, it should be correct.
You can find the difference in the `load` line
for `aten.argmax(x, 1)`
```
auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + **static_cast<long>(x1 + (144L*x0)), 16)**;
```
for `aten.argmax(x, 0)`
```
auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(x0 + (144L*x1)), 16);
```

Reply:
Ah, yes, my bad. Thanks for the detailed explanation!

Reply:
Oh, in fact, can we make the horizontal flag compile time, not runtime, like passing it as a template arg?